File: go/run_length_encoding.go
package encode

import (
	"fmt"
	"strings"
	"unicode"
)

func encode(ct int, c rune) string {
	if ct == 0 {
		return ""
	}
	if ct == 1 {
		return string(c)
	}
	return fmt.Sprintf("%d%c", ct, c)
}

func RunLengthEncode(input string) string {
	var last rune
	ct := 0
	s := ""
	for _, c := range input {
		if last == c {
			ct++
		} else {
			s += encode(ct, last)
			ct = 1
			last = c
		}
	}

	s += encode(ct, last)
	return s
}

func decode(ct int, c rune) string {
	if ct < 2 {
		return string(c)
	}
	return strings.Repeat(string(c), ct)
}

func RunLengthDecode(input string) string {
	ct := 0
	s := ""
	for _, c := range input {
		if unicode.IsDigit(c) {
			ct = ct*10 + int(c-'0')
		} else {
			s += decode(ct, c)
			ct = 0
		}
	}

	return s
}

File: dart/run_length_encoding.dart
class RunLengthEncoder {
  String encode(String input) {
    String s = '';
    int ct = 1;
    String pre = '';
    for (int i = 0; i < input.length; i++) {
      if (pre == input[i]) {
        ct++;
      } else {
        if (ct > 1) {
          s += '$ct';
        }
        s += '$pre';
        ct = 1;
        pre = input[i];
      }
    }
    if (ct > 1) {
      s += '$ct';
    }
    s += '$pre';
    return s;
  }

  String decode(String input) {
    int ct = 0;
    String s = '';
    for (int i = 0; i < input.length; i++) {
      final n = int.tryParse(input[i]);
      if (n == null) {
        while (ct > 1) {
          s += input[i];
          ct--;
        }
        s += input[i];
        ct = 0;
      } else {
        ct = ct * 10 + n;
      }
    }
    return s;
  }
}
